cluster-autoscaler/cloudprovider/mcm/mcm_manager.go

```diff
@@ -397,181 +401,196 @@ func (m *McmManager) Cleanup() {
+	// get the machine deployment and return if rolling update is not finished
+	md, err := m.machineDeploymentLister.MachineDeployments(m.namespace).Get(commonMachineDeployment.Name)
 	if err != nil {
-		klog.Errorf("Unable to fetch MachineDeployment object %s, Error: %s", commonMachineDeployment.Name, err)
+		klog.Errorf("Unable to fetch MachineDeployment object %s, Error: %v", commonMachineDeployment.Name, err)
 		return err
 	}
 	if !isRollingUpdateFinished(md) {
 		return fmt.Errorf("MachineDeployment %s is under rolling update , cannot reduce replica count", commonMachineDeployment.Name)
 	}
+	// update priorities of machines to be deleted except the ones already in termination to 1
+	scaleDownAmount, err := m.prioritizeMachinesForDeletion(targetMachineRefs, commonMachineDeployment.Name)
+	if err != nil {
+		return err
+	}
+	// Trying to update the machineDeployment till the deadline
+	err = m.retry(func(ctx context.Context) (bool, error) {
+		return m.scaleDownMachineDeployment(ctx, commonMachineDeployment.Name, scaleDownAmount)
+	}, "MachineDeployment", "update", commonMachineDeployment.Name)
+	if err != nil {
+		klog.Errorf("unable to scale in machine deployment %s, Error: %v", commonMachineDeployment.Name, err)
+		return fmt.Errorf("unable to scale in machine deployment %s, Error: %v", commonMachineDeployment.Name, err)
+	}
+	return nil
+}
 
-	for _, machine := range machines {
-
-		// Trying to update the priority of machine till retryDeadline
-		retryDeadline := time.Now().Add(maxRetryTimeout)
-		for {
-			machine, err := m.machineLister.Machines(m.namespace).Get(machine.Name)
-			if err != nil && time.Now().Before(retryDeadline) {
-				klog.Warningf("Unable to fetch Machine object %s, Error: %s , will retry in %s", machine.Name, err, conflictRetryInterval)
-				time.Sleep(conflictRetryInterval)
-				continue
-			} else if err != nil {
-				// Timeout occurred
-				klog.Errorf("Unable to fetch Machine object %s, Error: %s", machine.Name, err)
+// resetPriorityForNotToBeDeletedMachines resets the priority of machines with nodes without ToBeDeleted taint to 3
+func (m *McmManager) resetPriorityForNotToBeDeletedMachines(mdName string) error {
+	allMachinesForMachineDeployment, err := m.getMachinesForMachineDeployment(mdName)
+	if err != nil {
+		return fmt.Errorf("unable to list all machines for node group %s, Error: %v", mdName, err)
+	}
+	for _, machine := range allMachinesForMachineDeployment {
+		val, ok := machine.Annotations[machinePriorityAnnotation]
+		if ok && val != defaultPriorityValue {
+			if err := m.retry(func(ctx context.Context) (bool, error) {
+				nodeName := machine.Labels[v1alpha1.NodeLabelKey]
+				node, err := m.nodeLister.Get(nodeName)
+				if err != nil && !kube_errors.IsNotFound(err) {
+					return true, fmt.Errorf("unable to get Node object %s for machine %s, Error: %v", nodeName, machine.Name, err)
+				} else if err == nil && taints.HasToBeDeletedTaint(node) {
```

Comment:
The nil check is required because if the error is a `NotFound` error, then no node object will be present. I can take the `IsNotFound` check inside the if block and remove the nil check.